proposals/0167-loader-v4.md

```diff
@@ -0,0 +1,447 @@
+  the header size of their loader (account owner) respectively
+
+#### SetProgramLength
+
+- Instruction accounts:
+  - `[writable]` The program account to change the size of.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `2u32`
+  - `u32` The new size after the operation.
+- Behavior:
+  - Charge 32 + new_size_in_bytes / cpi_bytes_per_unit CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account
+  - Check that the status stored in the program account is
+  `NeverBeenDeployed` or `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Check that there are enough funds in the program account for rent
+  exemption of the new length,
+  otherwise throw `InsufficientFunds`
+  - Set the length of the program account to the requested new size plus
+  the header size
+
+#### WithdrawExcessLamports
+
+- Instruction accounts:
+  - `[writable]` The program account to withdraw from.
+  - `[signer]` The authority of the program.
+  - `[writable]` The recipient account.
+- Instruction data:
+  - Enum variant `3u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least three instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and recipient account do not alias,
+    otherwise throw `AccountBorrowFailed`
+  - Check that the recipient account is writable,
+  otherwise throw `InvalidArgument`
+  - Verify the program account
+  - Transfer lamports which are not needed for rent exemption from the
+  program account to the recipient account
+
+#### EraseAndWithdrawAllLamports
+
+- Instruction accounts:
+  - `[writable]` The program account to withdraw from.
+  - `[signer]` The authority of the program.
+  - `[writable]` The recipient account.
+- Instruction data:
+  - Enum variant `4u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least three instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Check that program account and recipient account do not alias,
+    otherwise throw `AccountBorrowFailed`
+  - Check that the recipient account is writable,
+  otherwise throw `InvalidArgument`
+  - Verify the program account
+  - Check that the status stored in the program account is
+  `NeverBeenDeployed` or `Retracted`,
+  otherwise throw `InvalidArgument`
+  - Set the length of the program account to 0 (removing the header too)
+  - Transfer all lamports from the program account to the recipient account
+
+#### Deploy
+
+- Instruction accounts:
+  - `[writable]` The program account to deploy.
+  - `[signer]` The authority of the program.
+- Instruction data:
+  - Enum variant `5u32`
+- Behavior:
+  - Charge 32 CUs
+  - Check there are at least two instruction accounts,
+    otherwise throw `NotEnoughAccountKeys`
+  - Verify the program account
+  - Check that the slot stored in the program account is not the current
```

Comment:
We only need this check if the program is `Retracted`. If the current slot is zero (in block 0) and the status is `NeverBeenDeployed` then the deploy should still be allowed.